doc/sphinxext/missing_references.py

```diff
@@ -0,0 +1,241 @@
+"""
+This is a sphinx extension to freeze your broken reference problems
+when using ``nitpicky = True``.
+
+The basic operation is:
+
+1. Add this extension to your ``conf.py`` extensions.
+2. Add ``missing_references_write_json = True`` to your ``conf.py``
+3. Run sphinx-build. It will generate ``missing-references.json``
+    next to your ``conf.py``.
+4. Remove ``missing_references_write_json = True`` from your
+    ``conf.py`` (or set it to ``False``)
+5. Run sphinx-build again, and ``nitpick_ignore`` will
+    contain all of the previously failed references.
+
+"""
+
+from collections import defaultdict
+import json
+import logging
+import os.path
+import posixpath
+
+from docutils.utils import get_source_line
+from docutils import nodes
+from sphinx.util import logging as sphinx_logging
+
+logger = sphinx_logging.getLogger(__name__)
+
+
+class MissingReferenceFilter(logging.Filter):
+    """
+    A logging filter designed to record missing reference warning messages
+    for use by this extension
+    """
+    def __init__(self, app):
+        self.app = app
+        super().__init__()
+
+    def _record_reference(self, record):
```

Comment:
Not sure why you need this separate method when it only has one caller, which is also a method on the same class.

Reply:
Its done as a convenience – the filter() method must return `True` or `False`, but the `_record_reference` method doesn't actually do any log filtering (all messages should get logged) so its a bit clearer IMO to return `True` from the filter method and put the logic somewhere else (so that we don't have to ensure that early returns from `_record_reference` also return `True`).